_posts/2023-04-21-extensibility-roadmap.markdown

```diff
@@ -0,0 +1,154 @@
+
+### Rewrite most plugins as extensions
+
+We will provide a line of sight for deprecating plugins by enabling extension migration with as close to one-to-one correspondence with plugin features as possible with the aim of making the effort required to migrate to extensions minimal. We will write a migration guide clearly specifying the needed effort to perform initial migration and follow-up deprecated feature replacement so that they can work it into their own task planning, offer code that makes the migration from plugins to extensions easier, implement samples that provide one-to-one analogues with the existing plugin framework, and create assurances that external behavior in the migration has not changed. Code permitting this quick bridge may be marked deprecated, but will allow users to methodically remove the deprecated code over time.
+
+This is [opensearch-sdk-java#315](https://github.com/opensearch-project/opensearch-sdk-java/issues/315).
+
+### Deprecate plugins, unbundle distributions
+
+Assuming extensions have been widely adopted, we can deprecate the plugin APIs to be removed in the next major version of OpenSearch. We don't expect this to happen earlier than OpenSearch 4.0. Older versions of plugins continue to work with older versions of OpenSearch and continue receiving security patches.
+
+We will replace the two distribution flavors of OpenSearch/Dashboards (currently a -min distribution without security or plugins) by a set of distribution manifests tailored for purposes such as log analytics vs. search. Each distribution will be represented by a manifest that can be assembled “on-the-fly” by downloading the artifacts from the public catalog for packaging purposes. We want to enable single-click install a collection of extensions, recommended distributions of OpenSearch for tailored purposes, ability for vendors to create their favorite flavor of OpenSearch/Dashboards distribution easily, and add the capability to create enterprise-tailored distributions.
+
+## Future
+
+Extensions make a lot of new, big ideas possible! Below are some of our favorites.
+
+### Hot swap extensions 
+
+OpenSearch/Dashboards bootstrap plugins at start time and various parts of the system assume that plugins do not change at runtime. Requiring a cluster restart for all extensions is a severely crippling limitation on the path of ecosystem adoption of any significant number of extensions, primarily because cluster restarts mean stopping inbound traffic. We ensure that any part of the system can be loaded or unloaded at runtime, starting with making settings in OpenSearch core dynamic, and add tooling to support loading and unloading extensions at runtime without restarting OpenSearch/Dashboards nodes or the entire cluster. This creates the ability to add, upgrade, or remove an extension without the need to restart OpenSearch/Dashboards, or connect a remote extension to an existing OpenSearch/Dashboards cluster.
```

Comment:
```suggestion
OpenSearch/Dashboards bootstrap plugins at start time and various parts of the system assume that plugins do not change at runtime. Requiring a cluster restart for all extensions is a severely crippling limitation on the path of ecosystem adoption of any significant number of extensions, primarily because cluster restarts mean stopping inbound traffic. We ensure that any extension point can be loaded or unloaded at runtime, starting with making settings in OpenSearch core dynamic, and add tooling to support loading and unloading extensions at runtime without restarting OpenSearch/Dashboards nodes or the entire cluster. This creates the ability to add, upgrade, or remove an extension without the need to restart OpenSearch/Dashboards, or connect a remote extension to an existing OpenSearch/Dashboards cluster.
```

Reply:
Taken!